// react-js-utl/hooks/useNestedDataCallback.js
function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

/*
 * Copyright (c) 2022 Anton Bagdatyev (Tonix)
 *
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without
 * restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following
 * conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */
import { useCallback } from "react";
import { shallowEqual } from "js-utl";
import ImmutableLinkedOrderedMap from "immutable-linked-ordered-map";
import useFactory from "./useFactory";
import usePrevious from "./usePrevious";
import declarativeFactory from "declarative-factory";
import visitor from "../primitives/visitor";
import isPrimitiveOfType from "../primitives/predicates/isPrimitiveOfType";

var dataGetKey = function dataGetKey(data, key) {
  return data.get(key);
};
/**
 * @type {Object}
 */


var mapTraverseFactory = {
  traverse: dataGetKey
};
/**
 * @type {Object}
 */

var weakMapTraverseFactory = {
  traverse: dataGetKey
};
/**
 * @type {Object}
 */

var immutableLinkedOrderedMapTraverseFactory = {
  traverse: dataGetKey
};
/**
 * @type {Object}
 */

var propTraverseFactory = {
  traverse: function traverse(data, key) {
    return data[key];
  }
};
/**
 * @type {Object}
 */

var yesVisitorKeyFactory = function yesVisitorKeyFactory(visitor) {
  return {
    key: function key() {
      return visitor.key;
    },
    visit: function visit() {
      return visitor.visit.apply(visitor, arguments);
    }
  };
};
/**
 * @type {Object}
 */


var noVisitorKeyFactory = function noVisitorKeyFactory(finalKey) {
  return {
    key: function key() {
      return finalKey;
    },
    visit: function visit() {}
  };
};
/**
 * Hook returning a callback to traverse nested data.
 *
 * @param {Array|Object|Map|WeakMap|ImmutableLinkedOrderedMap} data The data. Can be any of the specified types which in turn have nested data
 *                                                                  of any of the specified types.
 * @param {Object} [obj] An optional object with further parameters.
 * @param {boolean} [obj.pathCopy] Whether or not to create a copy of the current path and the data forming the path when traversing the nested data
 *                                 instead of pushing to a single path array.
 *                                 This parameter should be set to false only when the nested data is very deep and performance issues are experienced when using
 *                                 this hook for the given data.
 *                                 In case this parameter is set to false, eventual visitors (explained below) will receive an object where
 *                                 the "path" and "pathData" properties will refer to the same arrays for every visitor.
 * @return {(keys: Array) => *} The nested data callback which receives the keys and if called returns the nested data for the given keys, if any, or undefined.
 *                              The nested data callback takes an array of keys as argument. Note that the array is flattened (only its first dimension)
 *                              and therefore the following arrays will be treated as being the same array of keys:
 *
 *                                  import { useNestedDataCallback } from "react-js-utl/hooks";
 *
 *                                  // Inside functional component:
 *                                  useNestedDataCallback(data)(["a", "b", "c", "d", "e"]);
 *                                  useNestedDataCallback(data)(["a", "b", ["c", "d"], "e"]); // Same as above.
 *
 *                              Each element represents a nested key of the given data.
 *
 *                              An element can also be a visitor primitive returned by the "visitor" primitive function:
 *
 *                                  import { useNestedDataCallback } from "react-js-utl/hooks";
 *                                  import { visitor } from "react-js-utl/primitives";
 *
 *                                  // Inside functional component:
 *                                  useNestedDataCallback(data)([
 *                                      "a",
 *                                      visitor("b", ({
 *                                          currentData,
 *                                          depth,
 *                                          key,
 *                                          path,
 *                                          pathData,
 *                                          data,
 *                                      }) => {
 *                                          // Called when visiting the nested data ("currentData") for the "b" key (depth 1)
 *                                          // if the data for the "b" key is not "undefined".
 *                                      }),
 *                                      "c",
 *                                      "d",
 *                                      "e"
 *                                  ]);
 *
 *                              The visitor function allows to perform a custom behaviour when visiting the nested data.
 *                              Note that the visitor function will only be called if the underlying visited data for a given nested key
 *                              is not "undefined".
 */


export default function useNestedDataCallback(data) {
  var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
      _ref$pathCopy = _ref.pathCopy,
      pathCopy = _ref$pathCopy === void 0 ? true : _ref$pathCopy;

  var prevData = usePrevious(data);
  var finalDataFactory = useFactory(function () {
    return [[data instanceof Map || data instanceof WeakMap || ImmutableLinkedOrderedMap.isMap(data), function () {
      return prevData === data ? prevData : data;
    }], function () {
      return shallowEqual(prevData, data) ? prevData : data;
    }];
  }, [data]);
  var finalData = finalDataFactory();
  var callback = useCallback(function (keys) {
    var data = finalData;
    var effectiveKeys = keys.flat();
    var depth = 0;
    var currentData = data;
    var path = [];
    var pathData = [];

    var _iterator = _createForOfIteratorHelper(effectiveKeys),
        _step;

    try {
      var _loop = function _loop() {
        var key = _step.value;

        if (!currentData) {
          return "break";
        }

        var traverseFactory = declarativeFactory([[function () {
          return currentData instanceof Map;
        }, mapTraverseFactory], [function () {
          return currentData instanceof WeakMap;
        }, weakMapTraverseFactory], [function () {
          return ImmutableLinkedOrderedMap.isMap(currentData);
        }, immutableLinkedOrderedMapTraverseFactory], propTraverseFactory]);
        var visitorFactory = declarativeFactory([[function () {
          return isPrimitiveOfType(key, visitor);
        }, yesVisitorKeyFactory(key)], noVisitorKeyFactory(key)]);
        var finalKey = visitorFactory.key();
        currentData = traverseFactory.traverse(currentData, finalKey);

        if (pathCopy) {
          path = path.concat(finalKey);
          pathData = pathData.concat(currentData);
        } else {
          path.push(finalKey);
          pathData.push(currentData);
        }

        if (typeof currentData !== "undefined") {
          visitorFactory.visit({
            currentData: currentData,
            depth: depth,
            key: finalKey,
            path: path,
            pathData: pathData,
            data: data
          });
        }

        depth++;
      };

      for (_iterator.s(); !(_step = _iterator.n()).done;) {
        var _ret = _loop();

        if (_ret === "break") break;
      }
    } catch (err) {
      _iterator.e(err);
    } finally {
      _iterator.f();
    }

    return depth === effectiveKeys.length ? currentData : void 0;
  }, [finalData, pathCopy]);
  return callback;
}
